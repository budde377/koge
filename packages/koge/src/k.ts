type TemplateArgument = 
| string
| number
| boolean
| TemplateGenerator
| Array<TemplateArgument>

export type TemplateGenerateOptions = unknown

interface TemplateGenerator {
    generate(opts?: TemplateGenerateOptions): Generator<TemplateCommand>
}

export type Template = TemplateGenerator

type TemplateCommand = 
| {kind: 'string', value: string}
| {kind: 'number', value: number}
| {kind: 'boolean', value: boolean}

function *convertArgument(arg: TemplateArgument, options?: TemplateGenerateOptions): Generator<TemplateCommand> {
    if (typeof arg === 'string') {
        yield {kind: 'string', value: arg}
    } else if (typeof arg === 'number') {
        yield {kind: 'number', value: arg}
    } else if (typeof arg === 'boolean') {
        yield {kind: 'boolean', value: arg}
    } else if(Array.isArray(arg)) {
        for (const innerArg of arg) {
            yield* convertArgument(innerArg, options)
        }
    } else {
        yield* arg.generate(options)
    }

}

export function k(template: TemplateStringsArray, ...args: TemplateArgument[]): Template {
    return {
        *generate(options?: TemplateGenerateOptions) {
            for (let i = 0; i < template.length - 1; i ++) {
                yield {kind: 'string', value: template[i]}
                for (const command of convertArgument(args[i], options)) {
                    yield command
                }
            }
            yield {kind: 'string', value: template[template.length - 1]}
        }
    }
}
