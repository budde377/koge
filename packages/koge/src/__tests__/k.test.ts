import { k } from "../k"
import { mockPrinter } from "./utils"

describe('k', () => {
    it('it runs', () => {
        expect(mockPrinter(k`
        hello world
        ${true}
        ${1}
        ${1.2}
        ${1.}
        `)).toMatchSnapshot()
    })    
    it('supports nested templates', () => {
        const t = k`
            const v = 1234;
        `
        expect(mockPrinter(k`
            const a = 123;
            ${t}
        `)).toMatchSnapshot()
    })
    it('supports array', () => {
        expect(mockPrinter(k`
            ${[
                1234,
                k` foobar `,
                true,
                [
                    123,
                    ' ',
                    false
                ]
            ]}
        `)).toMatchSnapshot()
    })
})

