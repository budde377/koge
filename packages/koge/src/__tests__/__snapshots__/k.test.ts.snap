// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`k it runs 1`] = `
"
        hello world
        true
        1
        1.2
        1
        "
`;

exports[`k supports array 1`] = `
"
            1234 foobar true123 false
        "
`;

exports[`k supports nested templates 1`] = `
"
            const a = 123;
            
            const v = 1234;
        
        "
`;
